### PR TITLE
Feature/codegen feedback loop

### PR DESCRIPTION
**Pull request by morten-holm**

### CODE REVIEW

File: src/storm/codegen.ts [hunk 1]
```diff
@@ -13,22 +13,36 @@ import {
...
+import {readFileSync, writeFileSync} from 'fs';
 import YAML from "yaml";
+import * as fs from "node:fs";
+import {v4 as uuidv4} from 'uuid';
 
 type ImplementationGenerator = (prompt: StormFileImplementationPrompt, conversationId?: string) => Promise<StormStream>;
 
+interface ErrorClassification {
+    error: string;
+    lineNumber: number;
```
Comment (sorenmat) [inline src/storm/codegen.ts:42]: how does this work if an error spans multiple lines?

File: src/storm/codegen.ts [hunk 2]
```diff
@@ -405,53 +420,170 @@ export class StormCodegen {
...
+        const relevantFiles = allFiles.filter(file => file.type != AIFileTypes.IGNORE);
+
+        for (let attempts = 1;  attempts <= 5; attempts++) {
+            if (fileErrors.length == 0) {
+                console.log(`No more errors for ${filename}`);
+                return;
+            }
+
+            console.log(`Errors in ${filename} - requesting error details`);
+            const filesForContext = await this.getErrorDetailsForFile(basePath, filename, [fileErrors[0]], relevantFiles, language);
```
Comment (sorenmat) [inline src/storm/codegen.ts:454]: what's with the fileErrors[0]??

File: src/storm/codegen.ts [hunk 3]
```diff
@@ -405,53 +420,170 @@ export class StormCodegen {
...
+            const fix = this.createFixRequestForFile(basePath, filename, [fileErrors[0]], filesForContext, relevantFiles, language);
+            const codeFixFiles = await this.codeFix(fix);
+            console.log(`Got fixed code for ${filename}`);
+            for (const codeFixFile of codeFixFiles) {
+                const filePath = codeFixFile.filename.indexOf(basePath) > -1 ? codeFixFile.filename : join(basePath, codeFixFile.filename);
+                const existing = readFileSync(filePath);
+                if (existing.toString().replace(/(\r\n|\r|\n)+$/, '') == codeFixFile.content.replace(/(\r\n|\r|\n)+$/, '')) {
+                    console.log(`${filename} not changed by gemini`);
+                }
+                writeFileSync(filePath, codeFixFile.content);
```
Comment (sorenmat) [inline src/storm/codegen.ts:466]: I guess there is no need to write it if it hasn't changed?

File: src/storm/codegen.ts (hunk 1 above)
Comment (morten-holm) [reply src/storm/codegen.ts:42]: It's the line number from the compiler. The error from the compiler can span multiple lines and this is also true for the expression in the source file. We assume that the output from the compiler is right.

File: src/storm/codegen.ts (hunk 2 above)
Comment (morten-holm) [reply src/storm/codegen.ts:454]: One error at a time, so we concentrate on the first. Should probably not be an array - will fix.

**Pull request merged**